src/platforms/evdev/platform.cpp

```diff
@@ -427,9 +430,14 @@ void mie::Platform::device_added(libinput_device* dev)
 
     try
     {
-        devices.emplace_back(std::make_shared<mie::LibInputDevice>(report, std::move(device_ptr)));
+        auto new_device = std::make_shared<mie::LibInputDevice>(
+            report, led_observer_registrar, std::move(device_ptr));
+        devices.emplace_back(new_device);
 
-        input_device_registry->add_device(devices.back());
+        auto weak_device = input_device_registry->add_device(devices.back());
+
+        if (auto device = weak_device.lock())
+            new_device->associate_to_id(device->id());
```

Comment:
Oh!

So we're making `mi::Device` visible and passing `LedObserverRegistrar` around so that we can do two-phase initialization here?

Can we simplify things by registering the callback from `DefaultInputDeviceHub`/`DefaultDevice`?

Reply:
That would require `InputDevice` to have `set_leds` defined on it, which would break the input ABI again :smile: That's how I had it before.

Reply:
But I guess we broke it anyway my change the `create_platform` signature :joy: